pyrometheus/__init__.py

```diff
@@ -454,57 +514,51 @@ def get_temperature(self, H_or_E, T_guess, Y, do_energy=False):
 
         num_iter = 500
         tol = 1.0e-6
-        T_i = T_guess
-        dT = 1.0
-        F  = H_or_E
-        J  = 0.0
-
-        for iter in range( 0, num_iter ):
-            F    -= he_fun( T_i, Y )
-            J    -= pv_fun( T_i, Y )
-            dT    = - F / J
-            T_i  += dT
-            if np.abs( dT ) < tol:
-                T = T_i
+        ones = (1 + enthalpy_or_energy) - enthalpy_or_energy
+        t_i = t_guess * ones
+
+        for iter in range(num_iter):
+            f = enthalpy_or_energy - he_fun(t_i, y)
+            j = -pv_fun(t_i, y)
+            dt = -f / j
+            t_i += dt
+            if _pyro_norm(self.usr_np, dt, np.inf) < tol:
```

Comment:
There are a couple of things to consider that tie into this. An easy way to think about the root of the issue is that the output temperatures are sensitive to the initial guess passed to this routine.   The ideal situation is that we pass the actual/current temperature as the initial guess - we typically settle for a "reasonable" guess or for the temperature at the last time step.

If an initial guess is far off from the actual temperature, then this Newton iter can (a) take much more iters to converge to good temp, (b) converge to the wrong result or (c) fail to converge within num_iter. (a), (b) are the main things we want to avoid with restarting the DV (at least temperature for now).  Additionally, since the output temps are sensitive to input temps - we *need* to restart temperature in order to get deterministic results through restart - and this is important.

Unless we fix the temperature restart issues, we can expect that we will regularly encounter situation (c),  but because the starting guess and the actual temperature are not close.  In this case, we don't want to pass back NAN, we want to actually find the valid temperature, and so reverting to a more expensive procedure to find it would be a reasonable thing to do.   Or - until we get it fixed we can set num_iter to something insanely high.  Heh.

However if we *fix* the temperature restart issue and start actually using the current(ish) temperature as the initial guess, then if we could stop the code on encountering condition (c), it'd be terrific.  

Another potential fix is to roll a routine that provides a deterministic guess for a starting temperature based on the input state.   Then we'd only need to save state as we strongly desire, and we'd still get a deterministic temperature calc.   The big problem there is coming up with a sane temperature estimator routine; but it might be relatively easy to roll mechanism=specific estimators right into the generated code.

Reply:
> An easy way to think about the root of the issue is that the output temperatures are sensitive to the initial guess passed to this routine.

Just so I have this clear: What you're saying is that the equation being solved here doesn't suffice to fully define the temperature, i.e. the  initial guess matters? (If so, that seems strange to me.)

> Unless we fix the temperature restart issues, we can expect that we will regularly encounter situation (c), but because the starting guess and the actual temperature are not close.

Sure, Newton can fail. But wouldn't you rather know that it did than not?

A different aspect is what this dependency does to some of the interfaces we have in mirgecom. IIUC what you're saying, then you can't reasonably ask for temperatures without supplying a starting guess. That means the EOS interface in mirgecom needs to change.

(If you see this in time, could you add these discussions to the agenda for tomorrow?)